### PR TITLE
Add config flag to enable/disable remat HLO pass

### PR DESCRIPTION
**Pull request by keshavb96**
The flag --xla_disable_hlo_passes=rematerialization is often passed to XLA during compilation as the currently this pass often makes a poor choice of what to remat. With this patch the remat HLO pass is enabled by default but there is no need to pass --xla_disable_hlo_passes=rematerialization to disable it. Users can use the added config flag instead.

### CODE REVIEW

File: jax/_src/compiler.py [hunk 1]
```diff
@@ -53,6 +53,17 @@
     ),
 )
 
+_ENABLE_COMPILER_REMAT_OPTIMIZATION_PASS = config.bool_flag(
+    "jax_compiler_enable_remat_pass",
+    config.bool_env('JAX_COMPILER_ENABLE_REMAT_PASS', False),
```
Comment (yashk2810) [inline jax/_src/compiler.py:58]: You don't need `bool_env` anymore?

File: jax/_src/compiler.py [hunk 2]
```diff
@@ -53,6 +53,17 @@
     ),
 )
 
+_ENABLE_COMPILER_REMAT_OPTIMIZATION_PASS = config.bool_flag(
```
Comment (yashk2810) [inline jax/_src/compiler.py:56]: Use bool_state. See `config.py` to check how other flags are defined.

File: jax/_src/compiler.py (hunk 1 above)
Comment (yashk2810) [inline jax/_src/compiler.py:58]: default should be True, because that's the situation at HEAD IIUC. You can switch it to False for your use case.

**Comment by yashk2810** [comment]
Can you explain what problem you are solving? The current PR description does not explain that.

File: jax/_src/compiler.py (hunk 1 above)
Comment (nouiz) [reply jax/_src/compiler.py]: There was discussion with @hawkinsp that we want to change the default in JAX.
The current default is bad for a few reasons.
It isn't predictable as small change to the graph give wrong reasons.
Currently, we always disable it as it trigger too easily and does a poor selection of what to remat.

So the conclusion was to not enable it by default and document that users can enable it if they want to try it.

**Comment by keshavb96** [comment]
@yashk2810

File: jax/_src/config.py [hunk 3]
```diff
@@ -1528,6 +1528,14 @@ def _update_disable_jit_thread_local(val):
     default=True,
     help=('Enables using optimization-barrier op for lowering remat.'))
 
+enable_remat_opt_pass = bool_state(
+    name='jax_compiler_enable_remat_pass', 
+    default=False,
```
Comment (hawkinsp) [inline jax/_src/config.py:1533]: We certainly should check in this PR with the flag at its current default.

We can then explore, in a subsequent PR, flipping the default.

**Comment by keshavb96** [comment]
@hawkinsp should be set to the right default now

**Review by hawkinsp — APPROVED** [review]

**Pull request merged**